File: src/video_chunking.erl
-module(video_chunking).

-export([chunk_video/2, get_chunk/2]).

chunk_video(VideoPath, ChunkSize) ->
    case file:read_file(VideoPath) of
        {ok, VideoData} ->
            chunk_video_data(VideoData, ChunkSize, []);
        {error, Reason} ->
            error_logger:error_msg("Failed to read video file ~p: ~p~n", [VideoPath, Reason]),
            {error, Reason}
    end.

get_chunk(VideoPath, ChunkIndex) ->
    case file:read_file(VideoPath) of
        {ok, VideoData} ->
            Chunks = chunk_video_data(VideoData, 1048576, []),
            lists:nth(VideoIndex, Chunks);
        {error, Reason} ->
            error_logger:error_msg("Failed to read video file ~p: ~p~n", [VideoPath, Reason]),
            {error, Reason}
    end.

chunk_video_data(VideoData, ChunkSize, Acc) when byte_size(VideoData) =< ChunkSize ->
    lists:reverse([VideoData | Acc]);
chunk_video_data(VideoData, ChunkSize, Acc) ->
    <<Chunk:ChunkSize/binary, Rest/binary>> = VideoData,
    chunk_video_data(Rest, ChunkSize, [Chunk | Acc]).

File: src/scala/main/scala/video_chunking.scala
import java.io.{File, FileInputStream, IOException}
import java.nio.file.{Files, Paths}
import scala.util.{Try, Success, Failure}

object VideoChunking {

  def chunkVideo(videoPath: String, chunkSize: Int): Try[List[Array[Byte]]] = {
    Try {
      val videoFile = new File(videoPath)
      val videoData = Files.readAllBytes(videoFile.toPath)
      chunkVideoData(videoData, chunkSize, List.empty)
    }
  }

  def getChunk(videoPath: String, chunkIndex: Int): Try[Array[Byte]] = {
    chunkVideo(videoPath, 1048576).map { chunks =>
      if (chunkIndex >= 0 && chunkIndex < chunks.length) {
        chunks(chunkIndex)
      } else {
        throw new IndexOutOfBoundsException(s"Chunk index $chunkIndex out of bounds")
      }
    }
  }

  private def chunkVideoData(videoData: Array[Byte], chunkSize: Int, acc: List[Array[Byte]]): List[Array[Byte]] = {
    if (videoData.length <= chunkSize) {
      (videoData :: acc).reverse
    } else {
      val (chunk, rest) = videoData.splitAt(chunkSize)
      chunkVideoData(rest, chunkSize, chunk :: acc)
    }
  }
}
